virtcontainers/container.go

```diff
@@ -418,61 +418,64 @@ func newContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
 
 // createContainer creates and start a container inside a Pod. It has to be
 // called only when a new container, not known by the pod, has to be created.
-func createContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
+func createContainer(pod *Pod, contConfig ContainerConfig) (c *Container, err error) {
 	if pod == nil {
 		return nil, errNeedPod
 	}
 
-	c, err := newContainer(pod, contConfig)
+	c, err = newContainer(pod, contConfig)
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	if err := c.createContainersDirs(); err != nil {
-		return nil, err
+	if err = c.createContainersDirs(); err != nil {
+		return
 	}
 
-	if !c.pod.config.HypervisorConfig.DisableBlockDeviceUse {
-		agentCaps := c.pod.agent.capabilities()
-		hypervisorCaps := c.pod.hypervisor.capabilities()
-
-		if agentCaps.isBlockDeviceSupported() && hypervisorCaps.isBlockDeviceHotplugSupported() {
-			if err := c.hotplugDrive(); err != nil {
-				return nil, err
-			}
+	// In case the container creation fails, the following takes care
+	// of rolling back all the hotplug and mount operations.
+	defer func() {
+		if err != nil && c != nil {
+			c.removeResources()
+			c.detachDevices()
```

Comment:
If a device is shared with different containers in one pod, should we determine whether the device is out of use or just  hostunplug it.

Reply:
I don't know if we're trying to support this case yet.
But more than that, we're trying to rollback a container creation failure, the device is not yet shared with other containers, therefore I think it's not really related to this PR.

Reply:
Consider that  trying to create container C  in the pod, and use the `block device` like  `vdc` hotplugged by container B(concttainer A is `pause` contaienr ).  If we just hot unplug `vdc`, can it be a case that container B will face some problems even if it's an creation failure of container C.

If we finally decide not to support this case, maybe it's not really related to this PR.

Reply:
Ok so I have made some investigations and basically we don't completely support the case you're describing right now. I have opened an issue for that: https://github.com/kata-containers/runtime/issues/171

This is why this PR is not related to this issue since we don't even have the support for properly sharing the same device from inside the agent.

Reply:
And the roll back would not be an issue since we would detect that no device needs to be attached, meaning the `Container` structure will have no device associated, leading to the function `c.detachDevices()` to be a noop in case of container C.

Reply:
Thanks @sboeuf .  Since this is moving forward and fixing existing functionality, I'm okay with this.